### PR TITLE
examples/docker/ci: Assorted cleanups

### PR DESCRIPTION
**Pull request by phlax**
follow up from #25973 

- switch examples to new docker variant names
- refactor/improve/cleanup vrp recipe/tests
- better resolution to disk space issues (improve workaround)
- improved logging output
- docs cleanups
- consolidate/improve reuse of docker recipes
- reduce dependabot noise (by deduping deps)
- switch `flask` -> `aiohttp` to resolve CVEs and workaround buggy attitude/ehaviour (https://github.com/pallets/flask/issues/4938)

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
cc @htuch ive cleaned up the vrp build and testing - im pretty confident it does what is expected and now tests that also

File: examples/verify-common.sh [hunk 1]
```diff
@@ -31,6 +32,8 @@ bring_up_example_stack () {
     run_log "Bring up services ($path)"
     "${DOCKER_COMPOSE[@]}" "${up_args[@]}" || return 1
     echo
+    df -h
+    echo
```
Comment (phlax) [inline examples/verify-common.sh:36]: remove debugging

File: examples/vrp-local/verify.sh [hunk 2]
```diff
@@ -3,34 +3,52 @@
...
+test_vrp () {
+    wait_for 10 bash -c "responds_with 'normal' https://localhost:10000/content -k"
+    run_log "Test proxy ${1}"
+    responds_with \
+        normal \
+        https://localhost:10000/content \
+        -k
+}
+
+# Running this does not currently working with buildkit (even if that is how it is built).
```
Comment (phlax) [inline examples/vrp-local/verify.sh:37]:
```suggestion
# Running this is not currently working with buildkit (even if that is how it is built).
```

File: docs/root/start/sandboxes/setup.rst [hunk 3]
```diff
@@ -48,12 +48,18 @@ Install ``docker compose``
...
+.. note::
+   Any ``20.0+`` version of Docker provides a builtin ``docker compose`` command.
+
+   The sandboxes are tested using ``compose`` in this way, so this is preferable over using the python version.
 
 Docker Compose (``docker-compose``) can also be installed as a `python application <https://pypi.org/project/docker-compose/>`_ and can be
 installed through a variety of methods including `pip <https://pip.pypa.io/en/stable/>`_ and
-`native operating system installation <https://docs.docker.com/compose/install/>`_. Most of the sandboxes should also
-work using ``docker-compose``.
+`native operating system installation <https://docs.docker.com/compose/install/>`_.
```
Comment (phlax) [inline docs/root/start/sandboxes/setup.rst:60]: remove this

File: examples/shared/node/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
+FROM node:bullseye-slim as node-base
```
Comment (phlax) [inline examples/shared/node/Dockerfile:1]: add version/sha

**Comment by phlax** [comment]
ive had to replace flask with aiohttp in the cache example as flask is (again iirc) producing buggy/non-compliant responses and there attitude seems to be - dont use it

i will update all of the other examples to also use aiohttp in a follow up

File: examples/cache/verify.sh [hunk 5]
```diff
@@ -21,6 +21,7 @@ check_validated() {
     # Make sure there is no age header
     if grep -q "age:" <<< "$1"; then
         echo "ERROR: validated responses should not have an age header" >&2
+        echo "$1"
```
Comment (phlax) [inline examples/cache/verify.sh:24]: rm debug

File: examples/shared/python/aiohttp/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,2 @@
+aiohttp
```
Comment (phlax) [inline examples/shared/python/aiohttp/requirements.txt:1]: add requirements.in

File: examples/cache/service.py [hunk 7]
```diff
@@ -1,41 +1,88 @@
...
+
+
+def _set_etag_header(response) -> None:
+    etag = _compute_etag(response)
+    if etag is not None:
+        response.headers["Etag"] = etag
+
+
+@routes.get('/service/{service_number}/{response_id}')
+async def hello(request):
```
Comment (phlax) [inline examples/cache/service.py:60]:
```suggestion
async def get(request):
```

**Comment by phlax** [comment]
arggh, updating flask (to fix CVEs) breaks other examples - i will need to switch them also to use aiohttp i think

File: examples/cache/service.py [hunk 8]
```diff
@@ -1,41 +1,86 @@
...
+
+@routes.get('/service/{service_number}/{response_id}')
+async def get(request):
+    service_number = request.match_info["service_number"]
+    response_id = request.match_info["response_id"]
     stored_response = yaml.safe_load(open('/etc/responses.yaml', 'r')).get(response_id)
 
     if stored_response is None:
-        abort(404, 'No response found with the given id')
+        raise aiohttp.web.HTTPFound(reason="No response found with the given id")
```
Comment (phlax) [inline examples/cache/service.py:65]: not found

File: ci/do_ci.sh [hunk 9]
```diff
@@ -173,28 +173,38 @@ function run_process_test_result() {
...
+            filename="envoy.tar"
+        fi
+        oci_archive="${ENVOY_DOCKER_BUILD_DIR}/docker/${filename}"
+
+        echo "Copy oci image: oci-archive:${oci_archive} docker-daemon:envoyproxy/envoy:${variant}"
+        skopeo copy -q "oci-archive:${oci_archive}" "docker-daemon:envoyproxy/envoy:${variant}"
+        rm "$oci_archive"
+    done
+
+    docker info
```
Comment (phlax) [inline ci/do_ci.sh:196]: rm

File: ci/do_ci.sh [hunk 10]
```diff
@@ -173,28 +173,38 @@ function run_process_test_result() {
...
+
+    docker images | grep envoy
+
+    export DEBIAN_FRONTEND=noninteractive
+    sudo apt-get -qq update -y
+    sudo apt-get -qq install -y --no-install-recommends expect
+    export DOCKER_NO_PULL=1
+    export DOCKER_RMI_CLEANUP=1
+    umask 027
```
Comment (phlax) [inline ci/do_ci.sh:205]: add comment why this is here

File: examples/ext_authz/upstream/service/service.py [hunk 11]
```diff
@@ -1,12 +1,15 @@
...
-app = Flask(__name__)
+routes = web.RouteTableDef()
 
 
-@app.route('/service')
-def hello():
-    return 'Hello ' + request.headers.get('x-current-user') + ' from behind Envoy!'
+@routes.get("/service")
+async def get(request):
+    return web.json_response(
```
Comment (phlax) [inline examples/ext_authz/upstream/service/service.py:8]: `web.Response`

**Comment by phlax** [comment]
i have fully debugged the disk space issues while working on this PR

a few things are putting pressure on the disk - the docker build cache accumulates, the 2 compile/build jobs use a lot of /tmp space, and for some reason some images and containers dont seem to go away properly (eg mysql)

checking the build cache and there is definitely things that could be done more efficiently (combining go recipes eg) and there are some workarounds (clearing cache after build eg) but for the most part this is not the best approach

the compile jobs benefit from sharing the /tmp cache so it would be better not to remove these

in the end i just came to the conclusion that there is so much cruft that we might as well just delete (ie rm not apt-get remove) it - this can be quick and means we dont need to do any other workarounds

the ideal solution is to shift this to a self-hosted agent

**Comment by phlax** [comment]
@mattklein123 there are a couple more rounds of cleanups i want to do to the examples, but with this PR the bulk of the updates are out of the way

as mentioned elsewhere shifting this to a hosted agent would resolve quite a few issues (we need to update ci-infra for that)

in terms of CI my end goal here is to run these concurrently and incrementally (with bazel)  - i think we are not so far off that - and if we get it right it would reduce verify examples time to zero unless there are actual changes to either the binary or the example

**Comment by htuch** [comment]
@phlax have you manually verified VRP as per https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/google_vrp? I think it's fine otherwise. In theory VRP should be covered by CI but worth belts-and-braces if making a big change like this. 

Separately, I feel this is part of our security posture we aren't advertising well. Any thoughts on how to make this more visible?

**Comment by phlax** [comment]
> have you manually verified VRP as per ...

yep - it still does what is described there and now is tested

> this is part of our security posture we aren't advertising well

i have to admit - this is always the most problematic setup in the sandboxes, and my overriding thought is we need to make this more obvious or remove

> Any thoughts on how to make this more visible?

Add a proper sandbox.

The sandboxes here are end-user documentation and examples, that also have some integration tests - not the other way round

This has 2 sandboxes and no docs - not sure why that is, and the setup is not well thought through and a bit of a mess (does/not build/run with/out buildx)

We should consolidate into one and add documention - and _then_ make the test reflect what is documented.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**